ruma-events-macros/src/event_content.rs

```diff
@@ -7,42 +7,128 @@ use syn::{
+#[derive(Eq, PartialEq)]
 enum EventMeta {
     /// Variant holds the "m.whatever" event type.
     Type(LitStr),
+
+    /// Variant signals that this content type keeps redacted fields and is manually implemented.
+    SkipRedacted,
+
+    /// Variant signals when an event is a manually implemented redacted event.
+    RedactedCustom,
+}
+
+impl EventMeta {
+    fn is_event_type(&self) -> bool {
+        if let Self::Type(_) = self {
+            true
+        } else {
+            false
+        }
+    }
 }
 
 impl Parse for EventMeta {
     fn parse(input: ParseStream) -> syn::Result<Self> {
-        input.parse::<Token![type]>()?;
-        input.parse::<Token![=]>()?;
-        Ok(EventMeta::Type(input.parse::<LitStr>()?))
+        if input.parse::<Token![type]>().is_ok() {
+            input.parse::<Token![=]>()?;
+            Ok(EventMeta::Type(input.parse::<LitStr>()?))
+        } else if input.parse::<kw::skip_redacted>().is_ok() {
+            Ok(EventMeta::SkipRedacted)
+        } else if input.parse::<kw::custom_redacted>().is_ok() {
+            Ok(EventMeta::RedactedCustom)
+        } else {
+            Err(syn::Error::new(input.span(), "not a `ruma_event` attribute"))
+        }
     }
 }
 
 /// Create an `EventContent` implementation for a struct.
-pub fn expand_event_content(input: DeriveInput) -> syn::Result<TokenStream> {
+pub fn expand_event_content(input: &DeriveInput) -> syn::Result<TokenStream> {
     let ident = &input.ident;
 
-    let event_type_attr =
-        input.attrs.iter().find(|attr| attr.path.is_ident("ruma_event")).ok_or_else(|| {
+    let content_attr = input
+        .attrs
+        .iter()
+        .filter(|attr| attr.path.is_ident("ruma_event"))
+        .map(|attr| attr.parse_args::<EventMeta>())
+        .collect::<syn::Result<Vec<_>>>()?;
+
+    let event_type = {
+        let event_str = content_attr.iter().find(|a| a.is_event_type()).ok_or_else(|| {
             let msg = "no event type attribute found, \
             add `#[ruma_event(type = \"any.room.event\")]` \
             below the event content derive";
 
             syn::Error::new(Span::call_site(), msg)
         })?;
 
-    let event_type = {
-        let event_meta = event_type_attr.parse_args::<EventMeta>()?;
-        let EventMeta::Type(lit) = event_meta;
-        lit
+        if let EventMeta::Type(lit) = event_str {
+            lit
+        } else {
+            unreachable!("variant was checked to be of type EventMeta::Type")
+        }
+    };
+
+    let (redact_method, redacted) = if needs_redacted(input) {
+        let doc = format!("The payload for a redacted {}", ident);
+        let redacted_ident = quote::format_ident!("Redacted{}", ident);
+
+        (
+            quote! {
+                impl #ident {
+                    /// Transforms the full event content into a redacted content according to spec.
+                    pub fn redact(self) -> #redacted_ident {
+                        #redacted_ident
```

Comment:
This need to handle fields that don't get removed during redaction (those with `#[skip_redaction]`).

```suggestion
                        #redacted_ident { $(…)* }
```